headers/extract.py

```diff
@@ -118,76 +118,68 @@ def process_header(version, header):
+    # Remove exports defined with __attribute__ ((visibility ("default"))) (IL2CPPEXPORT/EXPORT)
+    # Basically iil2cpp_gc_wbarrier_set_field in current versions
+    header = re.sub(r'(?m)^.*?__attribute__ \(\(visibility \(\"default\"\)\)\).*$', '', header)
+    # Remove blank lines
+    header = re.sub(r'(?m)^[ \t]*\n', '', header)
+    # Change all class to struct
+    header = re.sub(r'(?m)^\s*class\s+', 'struct ', header)
+    # Change all C++ class inheritance to use composition for C compatibility
+    header = re.sub(r'(?m)^struct (.*?) : (?:public )?Il2Cpp(\S+)\s+\{([^\}]+)\};', r'struct \1 {\n Il2Cpp\2 \2;\3};', header)
     # Change all structs/enums/unions to use typedef for C compatibility
     header = re.sub(r'(?m)^(struct|union|enum)\s+(\w+)\s*\{([^\}]+(?:union\s*\{[^\}]+\}[^\}]*)*)^\};',
                     r'typedef \1 \2\n{\3} \2;', header)
     # Change all forward declarations into typedefs too
     header = re.sub(r'(?m)^struct(\s+)(\w+)\s*;', r'typedef struct\1\2 \2;', header)
+
     # Split the Il2CppClass structure into substructures so we can specialize certain fields
     # for e.g. vtable slot naming/typing
     m = re.search(r'struct\s+Il2CppClass\s*{\s*([^}]+)\n}(?:\s*Il2CppClass)?\s*;', header)
     if not m:
         raise Exception("struct Il2CppClass not found in %s" % version)
     header = header.replace(m.group(0), m.group(0) + '\n' + split_class_structure(version, m.group(1)))
-
-    # Add only the necessary bits from object-internals.h; this is stable across all versions
-    # (object-internals.h contains a lot of other C++ crap which would need to be stripped for IDA,
-    # and may pull in extra headers that would bloat our generated header)
-    header += '''
-struct MonitorData;
-struct Il2CppObject {
-    struct Il2CppClass *klass;
-    struct MonitorData *monitor;
-};
-typedef int32_t il2cpp_array_lower_bound_t;
-struct Il2CppArrayBounds {
-    il2cpp_array_size_t length;
-    il2cpp_array_lower_bound_t lower_bound;
-};
-struct Il2CppArray {
-    struct Il2CppObject obj;
-    struct Il2CppArrayBounds *bounds;
-    il2cpp_array_size_t max_length;
-    /* vector must be 8-byte aligned.
-       On 64-bit platforms, this happens naturally.
-       On 32-bit platforms, sizeof(obj)=8, sizeof(bounds)=4 and sizeof(max_length)=4 so it's also already aligned. */
-    void *vector[32];
-};
-struct Il2CppString {
-    struct Il2CppObject object;
-    int32_t length;
-    uint16_t chars[32];
-};
-'''
-
     return header
 
 for dir in glob.glob('../group*/il2cpp-*'):
     version = dir.split('il2cpp-')[1]
     dir = os.path.join(dir, 'libil2cpp')
-    fn = os.path.join(dir, 'class-internals.h')
+    fn = os.path.join(dir, 'object-internals.h')
     if not os.path.exists(fn):
-        fn = os.path.join(dir, 'il2cpp-class-internals.h')
+        fn = os.path.join(dir, 'il2cpp-object-internals.h')
+    args = ['cpp', '-P', '-D', 'NET_4_0', '-I', 'dummy', '-I', dir]
```

Comment:
I noticed that you define `NET_4_0`. Is this the default for Unity (even back to 5.3)? Just asking because I'm not sure of this myself.

Reply:
No, but the first use of this symbol in the header files we use is in Unity 5.5.0 and it became the default in Unity 2017.1 (I think) so that leaves only 5.5.0-5.6.7 where it is questionable.